### PR TITLE
feat(v2): createData plugin API

### PR DESCRIPTION
**Pull request by endiliey**
## Motivation

TL;DR: No changes in website functionality

`createModule` is a plugin actions api in `contentLoadeded` hooks that allows creating module.

Changes:
- Now global metadata such as `docsMetadata.json`, `blogMetadata.json` can be async imported instead of globally sync imported as context.
- We have less codes (cleaner @docusaurus/core)
- Sidebar now appears in SSR even when JavaScript is disabled.
Before
image

After
image

- Metadata is now namespaced to plugin
image

- Changed how headerlinks work
- We have way less bundle size now

Before: main bundle ~ 209 kb
after - 209kb

After: main bundle ~ 197kb
197 kb !!!

### Have you read the [Contributing Guidelines on pull requests](https://github.com/facebook/Docusaurus/blob/master/CONTRIBUTING.md#pull-requests)?

Yes

## Test Plan

- Locally nothing wrong
- Production (wait netlify)

### CODE REVIEW

File: packages/docusaurus-plugin-content-blog/src/index.js [hunk 1]
```diff
@@ -121,40 +119,46 @@ class DocusaurusPluginContentBlog {
...
+          JSON.stringify(metadataItem, null, 2),
+        );
+        if (isBlogPage) {
+          addRoute({
+            path: permalink,
+            component: blogPageComponent,
+            exact: true,
+            modules: {
+              entries: metadataItem.posts.map(post => ({
+                // To tell routes.js this is an import and not a nested object to recurse.
```
Comment (yangshun) [inline packages/docusaurus-plugin-content-blog/src/index.js]: Interesting. I had the exact same comment before :O

Comment (endiliey) [reply packages/docusaurus-plugin-content-blog/src/index.js]: Its indeed your comment. The diff on github wasn't perfect.

**Comment by endiliey** [comment]
I will address the review soon and modify the naming. The reasoning of having to write JSON.stringify is because I was thinking that maybe its not always a .json file, it can be a js file and they might want “export const” instead.

For further API extension, I agree it should be in later PR
